src/libraries/Microsoft.Extensions.Hosting/src/Internal/Host.cs

```diff
@@ -290,5 +294,26 @@ static async ValueTask DisposeAsync(object o)
                 }
             }
         }
+
+        private string DebuggerToString()
+        {
+            return $@"ApplicationName = ""{_hostEnvironment.ApplicationName}"", IsRunning = {(IsRunning ? "true" : "false")}";
+        }
+
+        // Host is running if the app has been started and hasn't been stopped.
+        private bool IsRunning => _applicationLifetime.ApplicationStarted.IsCancellationRequested && !_applicationLifetime.ApplicationStopped.IsCancellationRequested;
```

Comment:
Is this going to be misleading if stop has been requested but the host is still running because things are still shutting down?  I'd expect that to show IsRunning == true but it looks like it'll be IsRunning == false?

Reply:
I believe this has the behavior you want.

There is `ApplicationStarted`, `ApplicationStopping` and `ApplicationStopped`. This check uses `ApplicationStopped`. So if someone has requested a stop, then stopping is true, and IsRunning stays true until the stopped is true.

Reply:
Ok. I hadn't noticed a distinction between stopping and stopped.  Why in the code do we do:
```C#
_applicationLifetime.NotifyStopped();
try { await _hostLifetime.StopAsync(token).ConfigureAwait(false); }
catch (Exception ex) { exceptions.Add(ex); }
```
rather than:
```C#
try { await _hostLifetime.StopAsync(token).ConfigureAwait(false); }
catch (Exception ex) { exceptions.Add(ex); }
_applicationLifetime.NotifyStopped();
```
Has all work actually quiesced by this point so everything is indeed stopped and we're calling StopAsync again purely for it to propagate exceptions that have already occurred and been stored?

Reply:
In your code samples there is the application lifetime (have all the hosted services stopped) and the host lifetime (has the host stopped).

`IsRunning` is currently based on the application lifetime. The ConsoleLifetime literally doesn't do anything when stopping, but changing `IsRunning` to include the host lifetime would be a good change.

https://github.com/dotnet/runtime/blob/f25bc7b5b859935192e4eddc5b71efe71a60a4de/src/libraries/Microsoft.Extensions.Hosting/src/Internal/ConsoleLifetime.cs#L89-L93